### PR TITLE
Http server

### PR DESCRIPTION
**Pull request by TingDaoK**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/http/request_response.h [hunk 1]
```diff
@@ -319,8 +325,9 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
  * aws_http_stream_send_response() should be used to send a response.
  */
 AWS_HTTP_API
-struct aws_http_stream *aws_http_stream_new_server_request_handler(
-    const struct aws_http_request_handler_options *options);
+int aws_http_stream_configure_server_request_handler(
+    struct aws_http_stream *stream,
+    const struct aws_http_request_handler_options *options); // TODO
```
Comment (ColdenCullen) [inline include/aws/http/request_response.h]: What's `TODO`? Can you add more information, or remove if the task is complete?

File: include/aws/http/request_response.h [hunk 2]
```diff
@@ -351,7 +358,9 @@ int aws_http_stream_get_incoming_request_uri(const struct aws_http_stream *strea
 
 /* only callable from "request handler" streams */
 AWS_HTTP_API
-int aws_http_stream_send_response(struct aws_http_stream *stream, const struct aws_http_response_options *options);
+int aws_http_stream_send_response(
+    struct aws_http_stream *stream,
+    const struct aws_http_response_options *options); // TODO
```
Comment (ColdenCullen) [inline include/aws/http/request_response.h]: Same

File: source/h1_connection.c [hunk 3]
```diff
@@ -59,6 +59,9 @@ static size_t s_handler_initial_window_size(struct aws_channel_handler *handler)
 static size_t s_handler_message_overhead(struct aws_channel_handler *handler);
 static void s_handler_destroy(struct aws_channel_handler *handler);
 static struct aws_http_stream *s_new_client_request_stream(const struct aws_http_request_options *options);
+int s_configure_server_request_handler_stream(
```
Comment (ColdenCullen) [inline source/h1_connection.c]: This function can be `static`

File: source/h1_connection.c [hunk 4]
```diff
@@ -518,6 +522,73 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
     return NULL;
 }
 
+struct h1_stream *new_server_stream(struct h1_connection *connection) {
```
Comment (ColdenCullen) [inline source/h1_connection.c]: This function can be `static`, and we usually start static function names with `s_` (see other functions in this file as an example)

File: source/h1_connection.c [hunk 5]
```diff
@@ -518,6 +522,73 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+        AWS_FATAL_ASSERT(!err);
+        int new_stream_error_code = AWS_ERROR_SUCCESS;
+        if (connection->synced_data.new_stream_error_code) {
+            new_stream_error_code = connection->synced_data.new_stream_error_code;
+        } else {
+            aws_linked_list_push_back(&connection->synced_data.pending_stream_list, &stream->node);
+            /*if (!connection->synced_data.is_outgoing_stream_task_active) {
+                connection->synced_data.is_outgoing_stream_task_active = true;
+                should_schedule_task = true;
+            }*/
```
Comment (ColdenCullen) [inline source/h1_connection.c]: Delete this commented out code please

File: source/h1_connection.c [hunk 6]
```diff
@@ -518,6 +522,73 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+        (void *)stream,
+        (void *)options->server_connection,
+        AWS_BYTE_CURSOR_PRI(aws_http_version_to_str(options->server_connection->http_version)));
+
+    /*
+    if (should_schedule_task) {
+        AWS_LOGF_TRACE(AWS_LS_HTTP_CONNECTION, "id=%p: Scheduling outgoing stream task.", (void *)&connection->base);
+        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &connection->outgoing_stream_task);
+    }
+    */
```
Comment (ColdenCullen) [inline source/h1_connection.c]: Delete commented out code

File: source/h1_connection.c [hunk 7]
```diff
@@ -722,7 +793,7 @@ static void s_stream_complete(struct h1_stream *stream, int error_code) {
     struct h1_connection *connection = AWS_CONTAINER_OF(stream->base.owning_connection, struct h1_connection, base);
 
     /* Remove stream from list. */
-    aws_linked_list_remove(&stream->node);
+    aws_linked_list_remove(&stream->node); //? rmeove form where
```
Comment (ColdenCullen) [inline source/h1_connection.c]: What's this comment referring to?

File: source/h1_connection.c [hunk 8]
```diff
@@ -1035,8 +1106,22 @@ static int s_decoder_on_request(
     void *user_data) {
 
     struct h1_connection *connection = user_data;
+
+    AWS_ASSERT(!connection->thread_data.incoming_stream);
+    if (!connection->thread_data.incoming_stream) {
```
Comment (ColdenCullen) [inline source/h1_connection.c]: Why `if` if you're previously asserting?

File: source/h1_connection.c [hunk 9]
```diff
@@ -1035,8 +1106,22 @@ static int s_decoder_on_request(
...
+        /*make a new stream for this request and push it into the stream list wait for response
+        //only for server side!*/
```
Comment (ColdenCullen) [inline source/h1_connection.c]: You don't need `//` here

File: source/h1_connection.c [hunk 10]
```diff
@@ -1035,8 +1106,22 @@ static int s_decoder_on_request(
...
+        struct h1_stream *stream = new_server_stream(connection);
+        AWS_ASSERT(stream);
+        connection->base.server_data->on_incoming_request(
+            &(connection->base), &(stream->base), connection->base.server_data->connection_user_data);
```
Comment (ColdenCullen) [inline source/h1_connection.c]: You don't need to wrap these things in parens

File: source/h1_connection.c [hunk 11]
```diff
@@ -1493,15 +1580,22 @@ static int s_handler_process_read_message(
             /* Else processing message as normal HTTP data. */
             if (!connection->thread_data.incoming_stream) {
                 /* TODO: Server connection would create new request-handler stream at this point. */
+                if (connection->base.client_data)
+                /*client side*/
+                {
+                    if (connection->base.server_data) { /*error should never reach here*/
+                    }
```
Comment (ColdenCullen) [inline source/h1_connection.c]: You can remove this if statement

File: source/request_response.c [hunk 12]
```diff
@@ -317,6 +317,24 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
...
+int aws_http_stream_configure_server_request_handler(
+    struct aws_http_stream *stream,
+    const struct aws_http_request_handler_options *options) {
+    if (!options || options->self_size == 0 || !options->server_connection) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_CONNECTION,
+            "id=%p: Cannot create client request, options are invalid.",
+            (void *)(options ? options->server_connection : NULL));
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return AWS_OP_ERR;
```
Comment (ColdenCullen) [inline source/request_response.c]: `aws_raise_error` returns `AWS_OP_ERR`, so you can just write: `return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);`

File: tests/test_h1_client.c [hunk 13]
```diff
@@ -216,7 +216,7 @@ static enum aws_http_outgoing_body_state s_simple_send_body(
     void *user_data) {
 
     (void)stream;
-    struct simple_body_sender *data = user_data;
+    struct simple_body_sender *data = (struct simple_body_sender *)user_data;
```
Comment (ColdenCullen) [inline tests/test_h1_client.c]: This change shouldn't be necessary

File: tests/test_h1_server.c [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+} s_tester;
+
+static void s_test_on_complete(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct tester_request *request = user_data;
+    (void)stream;
+    struct aws_http_stream *r_handler = request->request_handler;
+    struct aws_byte_buf storage_buf = request->storage;
+
+    request->method = r_handler->incoming_request_method_str;
+    request->method.ptr = storage_buf.buffer + storage_buf.len;
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: Please move this line after the `write` call below, it's kind of confusing. Technically this doesn't point at valid memory until after the write completes.

File: tests/test_h1_server.c [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+    aws_byte_buf_write_from_whole_cursor(&storage_buf, r_handler->incoming_request_method_str);
+
+    request->uri = r_handler->incoming_request_uri;
+    request->uri.ptr = storage_buf.buffer + storage_buf.len;
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: Same

File: tests/test_h1_server.c [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+    aws_byte_buf_write_from_whole_cursor(&storage_buf, r_handler->incoming_request_uri);
+
+    // aws_http_stream_release(r_handler);
+    if (error_code == AWS_ERROR_SUCCESS) {
+    }
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: You can remove this

File: tests/test_h1_server.c [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+
+        my_header->value.ptr = storage->buffer + storage->len;
+        AWS_FATAL_ASSERT(aws_byte_buf_write_from_whole_cursor(storage, in_header->value));
+
+        in_header++;
+        my_header++;
+        request->num_headers++;
+    }
+    // int index = request->num_headers;
+    // request->headers[index];
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: Delete commented out code

File: tests/test_h1_server.c [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+    /* Copy data into storage, and point body cursor at that */
+    if (!request->body.ptr) {
+        request->body.ptr = request->storage.buffer + request->storage.len;
+    }
+    request->body.len += data->len;
+
+    if (request->stop_auto_window_update) {
+        *out_window_update_size = 0;
+    }
+    AWS_FATAL_ASSERT(aws_byte_buf_write_from_whole_cursor(&request->storage, *data));
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: Please move this write above the `request->body` modifications for reasons mentioned previously

File: tests/test_h1_server.c [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+    struct aws_http_request_handler_options options = AWS_HTTP_REQUEST_HANDLER_OPTIONS_INIT;
+    struct tester *test = user_data;
+
+    int index = test->request_num;
+    // initialize the new request
+    test->requests[index].num_headers = 0;
+    test->requests[index].has_incoming_body = false;
+    test->requests[index].header_done = false;
+
+    aws_byte_buf_init(&(test->requests[index].storage), test->alloc, 1024 * 1024 * 1);
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: Don't need parens around `test->requests[index].storage`

File: tests/test_h1_server.c [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+    options.server_connection = connection;
+    test->requests[index].request_handler = stream;
+    options.on_complete = s_test_on_complete;
+    options.on_request_headers = s_tester_on_request_header;
+    options.on_request_header_block_done = s_tester_on_request_header_block_done;
+    options.on_request_body = s_tester_on_request_body;
+
+    test->request_num++;
+    aws_http_stream_configure_server_request_handler(stream, &options);
+    // TODO
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: TODO.. what?

File: tests/test_h1_server.c [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+    aws_http_connection_release(s_tester.server_connection);
+    ASSERT_SUCCESS(testing_channel_clean_up(&s_tester.testing_channel));
+    aws_http_library_clean_up();
+    aws_logger_clean_up(&s_tester.logger);
+
+    return AWS_OP_SUCCESS;
+}
+
+/* For sending an aws_io_message into the channel, in the write or read direction */
+static int s_send_message_ex(struct aws_byte_cursor data) {
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: how about naming this `s_send_message_cursor`? `_ex` isn't descriptive

File: tests/test_h1_server.c [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+    ASSERT_NOT_NULL(msg);
+
+    ASSERT_TRUE(aws_byte_buf_write_from_whole_cursor(&msg->message_data, data));
+
+    ASSERT_SUCCESS(testing_channel_push_read_message(&s_tester.testing_channel, msg));
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_send_request_str(const char *str) {
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: How about `s_send_message_c_str`? `send_request` implies it does something different than `send_message`.

File: tests/test_h1_server.c [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+    ASSERT_TRUE(aws_byte_cursor_eq_c_str(&request.body, "write more tests"));
+
+    request = s_tester.requests[1];
+    ASSERT_TRUE(request.num_headers == 0);
+    ASSERT_TRUE(aws_byte_cursor_eq_c_str(&request.method, "GET"));
+    ASSERT_TRUE(aws_byte_cursor_eq_c_str(&request.uri, "/"));
+    /* clean up */
+    ASSERT_SUCCESS(s_server_tester_clean_up());
+    return AWS_OP_SUCCESS;
+}
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: need a newline at the end of the file

File: tests/test_h1_client.c (hunk 13 above)
Comment (TingDaoK) [reply tests/test_h1_client.c]: Oh, that is just because the VScode give me an error on that...

File: source/request_response.c [hunk 24]
```diff
@@ -317,6 +317,23 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
...
 }
 
+int aws_http_stream_configure_server_request_handler(
+    struct aws_http_stream *stream,
+    const struct aws_http_request_handler_options *options) {
+    if (!options || options->self_size == 0 || !options->server_connection) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_CONNECTION,
+            "id=%p: Cannot create client request, options are invalid.",
+            (void *)(options ? options->server_connection : NULL));
```
Comment (graebm) [inline source/request_response.c]: "...Cannot configure server stream..."

Also, use the `AWS_LS_HTTP_STREAM` identifier and pass the pointer to the stream, not the pointer to the connection

File: source/request_response.c [hunk 25]
```diff
@@ -317,6 +317,23 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
...
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    /* Connection owns stream, and must outlive stream */
+    aws_atomic_fetch_add(&options->server_connection->refcount, 1);
```
Comment (graebm) [inline source/request_response.c]: don't need to touch the refcount here

File: source/h1_connection.c [hunk 26]
```diff
@@ -518,6 +522,64 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+    if (!stream) {
+        /*log the bug */
+        return NULL;
+    }
+    int err;
+    { /* BEGIN CRITICAL SECTION */
+        err = aws_mutex_lock(&connection->synced_data.lock);
+        AWS_FATAL_ASSERT(!err);
+        int new_stream_error_code = AWS_ERROR_SUCCESS;
+        if (connection->synced_data.new_stream_error_code) {
```
Comment (graebm) [inline source/h1_connection.c]: the server stream is always created on the event-loop/channel/connection thread
so we should put the new stream in `thread_data.stream_list`, not `synced_data.pending_stream_list`
and we don't need to acquire the lock to do so

File: source/request_response.c [hunk 27]
```diff
@@ -317,6 +317,23 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
     return stream;
 }
 
+int aws_http_stream_configure_server_request_handler(
+    struct aws_http_stream *stream,
+    const struct aws_http_request_handler_options *options) {
+    if (!options || options->self_size == 0 || !options->server_connection) {
```
Comment (graebm) [inline source/request_response.c]: You should also check that this function is only called once ever, `aws_http_connection_configure_server()` does a similar check

File: include/aws/http/request_response.h [hunk 28]
```diff
@@ -319,7 +352,8 @@ struct aws_http_stream *aws_http_stream_new_client_request(const struct aws_http
  * aws_http_stream_send_response() should be used to send a response.
  */
 AWS_HTTP_API
-struct aws_http_stream *aws_http_stream_new_server_request_handler(
+int aws_http_stream_configure_server_request_handler(
```
Comment (graebm) [inline include/aws/http/request_response.h]: Update comments on this function

```
/**
 * Configure a server connection's new "request handler" stream.
 * This MUST be called from a server's on_incoming_request callback.
 */
```

File: source/h1_connection.c [hunk 29]
```diff
@@ -1035,8 +1097,20 @@ static int s_decoder_on_request(
...
     struct h1_connection *connection = user_data;
+
+    AWS_ASSERT(!connection->thread_data.incoming_stream);
+    /*make a new stream for this request and push it into the stream list wait for response
+      only for server side!*/
+    struct h1_stream *stream = s_new_server_stream(connection);
+    AWS_ASSERT(stream);
+    connection->base.server_data->on_incoming_request(
+        &connection->base, &stream->base, connection->base.server_data->connection_user_data);
+    connection->thread_data.incoming_stream = stream;
```
Comment (graebm) [inline source/h1_connection.c]: We should check to ensure that the user called configure(), and handle the errors if they didn't

File: include/aws/http/server.h [hunk 30]
```diff
@@ -129,7 +133,7 @@ struct aws_http_server_connection_options {
     /**
      * Invoked at the start of an incoming request.
      * Required.
-     * From this callback, user must call aws_http_create_request_handler().
+     * From this callback, user must call aws_http_stream_configure_server_request_handler().
      */
```
Comment (graebm) [inline include/aws/http/server.h]:
```
* Invoked when a new "request handler" stream is created to handle an incoming request.
* Required.
* From this callback, the user must call aws_http_stream_configure_server_request_handler().
* The user must call aws_stream_release() on the stream when they are done with it or its memory will never be cleaned up.
```

File: source/request_response.c (hunk 25 above)
Comment (TingDaoK) [reply source/request_response.c]: This will be fine

File: include/aws/http/private/request_response_impl.h [hunk 31]
```diff
@@ -41,9 +41,11 @@ struct aws_http_stream {
     aws_http_on_incoming_headers_fn *on_incoming_headers;
     aws_http_on_incoming_header_block_done_fn *on_incoming_header_block_done;
     aws_http_on_incoming_body_fn *on_incoming_body;
+    aws_http_on_request_end_fn *on_request_end;
     aws_http_on_stream_complete_fn *on_complete;
 
     struct aws_atomic_var refcount;
+    bool configured;
```
Comment (ColdenCullen) [inline include/aws/http/private/request_response_impl.h]: Could you be more specific? What was configured?

File: source/h1_connection.c [hunk 32]
```diff
@@ -200,6 +205,15 @@ struct h1_stream {
 
     /* Buffer for incoming data that needs to stick around. */
     struct aws_byte_buf incoming_storage_buf;
+
+    bool has_response;
```
Comment (ColdenCullen) [inline source/h1_connection.c]: This should go away, right?

File: source/h1_connection.c [hunk 33]
```diff
@@ -522,70 +542,162 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+        return AWS_OP_ERR;
+    }
+    /* valid status must be three digital code, change it into byte_cursor */
+    char status_code_str[4] = "XXX";
+    status_code_str[0] = (options->status / 100) % 10 + '0';
+    status_code_str[1] = ((options->status % 100) / 10) % 10 + '0';
+    status_code_str[2] = options->status % 10 + '0';
+    struct aws_byte_cursor status_code = aws_byte_cursor_from_c_str(status_code_str);
 
+    /* just for my own debug, deleted after test! */
```
Comment (ColdenCullen) [inline source/h1_connection.c]: ;)

File: source/h1_connection.c [hunk 34]
```diff
@@ -1593,15 +1724,42 @@ static int s_handler_process_read_message(
...
+                    /* Server side
+                     * Make a new stream for this request and push it into the stream list wait for response */
+                    struct h1_stream *stream = s_new_stream(&connection->base);
+                    if (!stream) {
+                        AWS_LOGF_ERROR(
+                            AWS_LS_HTTP_CONNECTION,
+                            "id=%p: Failed to create a new stream in connection. Closing connection.",
+                            (void *)&connection->base);
+                        goto shutdown;
+                        ;
```
Comment (ColdenCullen) [inline source/h1_connection.c]: remove

File: source/h1_connection.c [hunk 35]
```diff
@@ -1593,15 +1724,42 @@ static int s_handler_process_read_message(
...
+                        &connection->base, &stream->base, connection->base.server_data->connection_user_data);
+
+                    if (!stream->base.configured) {
+                        AWS_LOGF_ERROR(
+                            AWS_LS_HTTP_CONNECTION,
+                            "id=%p: The request handler stream is not configured,"
+                            "please make sure call aws_http_stream_configure_server_request_handler",
+                            (void *)&connection->base);
+                        goto shutdown;
+                        ;
```
Comment (ColdenCullen) [inline source/h1_connection.c]: remove

File: source/h1_connection.c [hunk 36]
```diff
@@ -518,6 +542,165 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+
+        err = aws_mutex_unlock(&connection->synced_data.lock);
+        AWS_FATAL_ASSERT(!err);
+    } /* END CRITICAL SECTION */
+    err = s_write_response_to_stream_buffer(h1_stream, options);
+    AWS_FATAL_ASSERT(!err);
+    h1_stream->base.stream_outgoing_body = options->stream_outgoing_body;
+    /* Success! */
+    AWS_LOGF_DEBUG(
+        AWS_LS_HTTP_STREAM, "id=%p: Created response on connection=%p: ", (void *)&h1_stream->base, (void *)connection);
```
Comment (ColdenCullen) [inline source/h1_connection.c]: `&h1_stream->base` can just be `stream`

File: source/h1_connection.c [hunk 37]
```diff
@@ -518,6 +542,165 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
...
+
+static int s_stream_send_response(struct aws_http_stream *stream, const struct aws_http_response_options *options) {
+    struct h1_stream *h1_stream = AWS_CONTAINER_OF(stream, struct h1_stream, base);
+    struct h1_connection *connection = AWS_CONTAINER_OF(stream->owning_connection, struct h1_connection, base);
+    int err;
+    bool should_schedule_task = false;
+    { /* BEGIN CRITICAL SECTION */
+        err = aws_mutex_lock(&connection->synced_data.lock);
+        if (h1_stream->synced_data.has_response) {
+            AWS_LOGF_ERROR(AWS_LS_HTTP_CONNECTION, "id=%p: Response already created on the stream", (void *)h1_stream);
```
Comment (ColdenCullen) [inline source/h1_connection.c]: Other places we're logging `id` as the `aws_http_stream *`, which in this case is `stream`, not `h1_stream`.

File: tests/test_h1_server.c [hunk 38]
```diff
@@ -0,0 +1,485 @@
...
+
+#if _MSC_VER
+#    pragma warning(disable : 4204) /* non-constant aggregate initializer */
+#endif
+
+#define TEST_CASE(NAME)                                                                                                \
+    AWS_TEST_CASE(NAME, s_test_##NAME);                                                                                \
+    static int s_test_##NAME(struct aws_allocator *allocator, void *ctx)
+
+/* get the request from the channel */
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: This comment doesn't make sense here

File: tests/test_h1_server.c [hunk 39]
```diff
@@ -0,0 +1,485 @@
...
+    if (!request->body.ptr) {
+        request->body.ptr = request->storage.buffer + request->storage.len - request->body.len;
+    }
+
+    if (request->stop_auto_window_update) {
+        *out_window_update_size = 0;
+    }
+}
+
+// create a new request handler
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: please use /* */ style comments

File: tests/test_h1_server.c [hunk 40]
```diff
@@ -0,0 +1,485 @@
...
+static void s_tester_on_incoming_request(
+    struct aws_http_connection *connection,
+    struct aws_http_stream *stream,
+    void *user_data) {
+    (void)connection;
+    struct aws_http_request_handler_options options = AWS_HTTP_REQUEST_HANDLER_OPTIONS_INIT;
+    struct tester *test = user_data;
+
+    int index = test->request_num;
+    // initialize the new request
```
Comment (ColdenCullen) [inline tests/test_h1_server.c]: /* */

File: tests/test_h1_server.c (hunk 23 above)
Comment (ColdenCullen) [reply tests/test_h1_server.c]: Still need a newline here

File: tests/test_h1_server.c (hunk 40 above)
Comment (ColdenCullen) [reply tests/test_h1_server.c]: Please use c style comments

File: tests/test_h1_server.c [hunk 41]
```diff
@@ -818,7 +800,6 @@ TEST_CASE(h1_server_send_multiple_responses_out_of_order_only_one_sent) {
     ASSERT_TRUE(request1->on_complete_error_code == AWS_ERROR_SUCCESS);
     /* last two failed, response 2 is missing */
     ASSERT_TRUE(request2->on_complete_error_code == AWS_ERROR_HTTP_CONNECTION_CLOSED);
-    /* do we need a new error code for this situation? */
     ASSERT_TRUE(request3->on_complete_error_code == AWS_ERROR_HTTP_CONNECTION_CLOSED);
```
Comment (TingDaoK) [inline tests/test_h1_server.c]: Do we need a new error code for this situation to inform our user?

File: source/h1_connection.c [hunk 42]
```diff
@@ -1215,8 +1443,18 @@ static int s_decoder_on_done(void *user_data) {
     }
 
     incoming_stream->is_incoming_message_done = true;
+    if (connection->base.server_data) {
+
+        if (incoming_stream->base.on_request_end) {
+            incoming_stream->base.on_request_end(&incoming_stream->base, incoming_stream->base.user_data);
+        }
+        if (incoming_stream->outgoing_state == STREAM_OUTGOING_STATE_DONE) {
+            AWS_ASSERT(&incoming_stream->node == aws_linked_list_begin(&connection->thread_data.stream_list));
```
Comment (TingDaoK) [inline source/h1_connection.c]: This may be useless

File: source/h1_connection.c [hunk 43]
```diff
@@ -1492,22 +1732,51 @@ static int s_handler_process_read_message(
...
+                        "connection.",
+                        (void *)&connection->base);
+
+                    aws_raise_error(AWS_ERROR_INVALID_STATE);
+                    goto shutdown;
+                } else {
+                    /* Server side
+                     * Make a new stream for this request and push it into the waiting stream list wait for response */
+                    struct h1_stream *stream = s_new_stream(&connection->base);
+                    if (!stream) {
```
Comment (TingDaoK) [inline source/h1_connection.c]: This will be caused by the allocator, maybe we do not log the error here.

File: source/h1_connection.c [hunk 44]
```diff
@@ -1672,6 +1941,11 @@ static int s_handler_shutdown(
             s_stream_complete(AWS_CONTAINER_OF(node, struct h1_stream, node), stream_error_code);
         }
 
+        while (!aws_linked_list_empty(&connection->thread_data.waiting_stream_list)) {
```
Comment (TingDaoK) [inline source/h1_connection.c]: If we need a new error code to inform user the circumstance that the stream already has a response, but those ones in front of it are not ready, and cause it failed to send the response

File: source/h1_connection.c [hunk 45]
```diff
@@ -1492,22 +1743,51 @@ static int s_handler_process_read_message(
...
+                            "please make sure call aws_http_stream_configure_server_request_handler",
+                            (void *)&connection->base);
+                        goto shutdown;
+                    }
+                    /* Waiting for response. */
+                    aws_linked_list_push_back(&connection->thread_data.waiting_stream_list, &stream->node);
 
-                aws_raise_error(AWS_ERROR_INVALID_STATE);
-                goto shutdown;
+                    connection->thread_data.incoming_stream = stream;
```
[the remark this answers is not in the document]
Comment (TingDaoK) [reply source/h1_connection.c]: Yes, there are only two places, where the s_update_incoming_stream_ptr() are called, and I made a if statement to make sure it is on the client side.

File: source/h1_connection.c [hunk 46]
```diff
@@ -1682,6 +1693,39 @@ static void s_connection_try_send_read_messages(struct h1_connection *connection
...
+     * If not, raise an error */
+    if (!stream->base.request_handler_configured) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_CONNECTION,
+            "id=%p: The request handler stream is not request_handler_configured,"
+            "please make sure call aws_http_stream_configure_server_request_handler",
+            (void *)&connection->base);
+        /* no need to release the stream here
+         * the configuration is not call the ref count is still 0 */
+        return NULL;
```
Comment (graebm) [inline source/h1_connection.c]: Destroy the stream if the user failed to configure it

**Pull request merged**